File: native/windows/src/ScreenFix.Core/Calibration/CalibrationLayout.cs
using ScreenFix.Core.Geometry;

namespace ScreenFix.Core.Calibration;

public readonly record struct CalibrationColor(string Hex, double Opacity = 1);

public sealed record CalibrationLayoutSpec(
    RectD SaveButton,
    RectD CancelButton,
    RectD Instruction,
    RectD InstructionDot,
    RectD InstructionText,
    double ButtonCornerRadius,
    double ButtonFontSize,
    double InstructionCornerRadius,
    double InstructionFontSize,
    CalibrationColor SaveColor,
    CalibrationColor CancelColor,
    CalibrationColor InstructionColor,
    CalibrationColor InstructionStrokeColor,
    CalibrationColor InstructionDotColor);

public readonly record struct CalibrationLayoutResult(
    CalibrationLayoutSpec? Value,
    string? Error)
{
    public bool IsSuccess => Value is not null;
}

public static class CalibrationLayout
{
    private const double MinimumWidth = 260;
    private const double MinimumHeight = 180;
    private const double Inset = 24;
    private const double ButtonGap = 12;
    private const double ButtonHeight = 42;

    public static CalibrationLayoutResult TryCreate(double width, double height)
    {
        if (!IsFinite(width) || !IsFinite(height) ||
            width < MinimumWidth || height < MinimumHeight)
        {
            return new CalibrationLayoutResult(
                null,
                "display is too small for calibration controls");
        }

        var buttonWidth = Math.Min(104, Math.Floor((width - (2 * Inset) - ButtonGap) / 2));
        var buttonY = height - Inset - ButtonHeight;
        var save = new RectD(Inset, buttonY, buttonWidth, ButtonHeight);
        var cancel = new RectD(Inset + buttonWidth + ButtonGap, buttonY, buttonWidth, ButtonHeight);
        var isNarrow = width < 378;
        var instructionHeight = isNarrow ? 58 : 42;
        var instructionWidth = isNarrow ? width - (2 * Inset) : 330;
        var instruction = new RectD(Inset, Inset, instructionWidth, instructionHeight);
        var dot = new RectD(40, Inset + ((instructionHeight - 8) / 2), 8, 8);
        var text = new RectD(
            58,
            Inset,
            isNarrow ? width - 98 : 280,
            instructionHeight);

        return new CalibrationLayoutResult(
            new CalibrationLayoutSpec(
                save,
                cancel,
                instruction,
                dot,
                text,
                9,
                16,
                10,
                isNarrow ? 13 : 15,
                new CalibrationColor("#16A34A"),
                new CalibrationColor("#353A42"),
                new CalibrationColor("#000000", 0.88),
                new CalibrationColor("#FFFFFF", 0.28),
                new CalibrationColor("#FF643B")),
            null);
    }

    private static bool IsFinite(double value) => !double.IsNaN(value) && !double.IsInfinity(value);
}

File: native/macos/Sources/ScreenFixCore/CalibrationLayout.swift
import Foundation

public struct CalibrationControlLayout: Equatable {
    public let save: RectD
    public let cancel: RectD
    public let instruction: RectD
    public let instructionDot: RectD
    public let instructionText: RectD
    public let instructionTextSize: Double

    public init(
        save: RectD,
        cancel: RectD,
        instruction: RectD,
        instructionDot: RectD,
        instructionText: RectD,
        instructionTextSize: Double
    ) {
        self.save = save
        self.cancel = cancel
        self.instruction = instruction
        self.instructionDot = instructionDot
        self.instructionText = instructionText
        self.instructionTextSize = instructionTextSize
    }

    public static func make(
        width: Double,
        height: Double,
        safeFrame: RectD? = nil
    ) -> CalibrationControlLayout? {
        guard width.isFinite, height.isFinite, width >= 260, height >= 180 else {
            return nil
        }
        let safe = safeFrame ?? RectD(x: 0, y: 0, width: width, height: height)
        guard isValid(safeFrame: safe, width: width, height: height) else { return nil }

        let buttonWidth = min(104, floor((safe.width - 48 - 12) / 2))
        let save = RectD(x: safe.x + 24, y: safe.bottom - 66, width: buttonWidth, height: 42)
        let cancel = RectD(x: safe.x + 36 + buttonWidth, y: safe.bottom - 66, width: buttonWidth, height: 42)
        if safe.width < 378 {
            let instruction = RectD(x: safe.x + 24, y: safe.y + 24, width: safe.width - 48, height: 58)
            return CalibrationControlLayout(
                save: save,
                cancel: cancel,
                instruction: instruction,
                instructionDot: RectD(x: instruction.x + 16, y: instruction.y + 25, width: 8, height: 8),
                instructionText: RectD(
                    x: instruction.x + 34,
                    y: instruction.y,
                    width: instruction.width - 50,
                    height: 58
                ),
                instructionTextSize: 13
            )
        }
        let instruction = RectD(x: safe.x + 24, y: safe.y + 24, width: 330, height: 42)
        return CalibrationControlLayout(
            save: save,
            cancel: cancel,
            instruction: instruction,
            instructionDot: RectD(x: instruction.x + 16, y: instruction.y + 17, width: 8, height: 8),
            instructionText: RectD(x: instruction.x + 34, y: instruction.y, width: 280, height: 42),
            instructionTextSize: 15
        )
    }

    private static func isValid(safeFrame: RectD, width: Double, height: Double) -> Bool {
        safeFrame.x.isFinite && safeFrame.y.isFinite
            && safeFrame.width.isFinite && safeFrame.height.isFinite
            && safeFrame.x >= 0 && safeFrame.y >= 0
            && safeFrame.width >= 260 && safeFrame.height >= 180
            && safeFrame.right <= width && safeFrame.bottom <= height
    }
}
